README.md
# vite-plugin-remix-routes

Use [Remix](https://github.com/remix-run/remix) routing in your [Vite](https://github.com/vitejs/vite) project.

## Plugin config

```js
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import remixRoutes from "vite-plugin-remix-routes";

export default defineConfig({
  plugins: [react(), remixRoutes()],
});
```

**With options:**

```js
remixRoutes({
  /* options here */
});
```

## Options

### appDirectory

https://remix.run/docs/en/v1/api/conventions#appdirectory

- **Optional**
- **Type**: `string`
- **Default**: `path.join(process.cwd(), "app")`

An absolute path to the folder containing the `routes` folder.

### dataRouterCompatible

- **Optional**
- **Type**: `boolean`
- **Default**: `true`

Set this to `false` if you're not using a router compatible with the data APIs
released in `react-router` 6.4.\
https://reactrouter.com/en/main/routers/picking-a-router#using-v64-data-apis

### importMode

> NOTE: This option only works if `dataRouterCompatible` is set to `false`.

- **Optional**
- **Type**: `(route: Route) => "async" | "sync"`
- **Default**: `() => "sync"`

A function that receives a `Route` to determine if the route's component should be imported synchronously or asynchronously.

### is404Route

> NOTE: This option only works if `dataRouterCompatible` is set to `false`. You should use an `ErrorBoundary` component instead.

- **Optional**
- **Type**: `(route: Route) => boolean`
- **Default**: `(route) => route.id.endsWith("/404")`

A function that receives a `Route` to determine if it should be a 404 route. (`path="*"`)

By default this matches all routes whose id's end with `/404`.

A route's `id` is the component path without extension.

### routes

https://remix.run/docs/en/v1/api/conventions#routes

- **Optional**
- **Type**: `(defineRoutes: DefineRoutesFunction) => Promise<ReturnType<DefineRoutesFunction>>`

A function for defining custom routes, in addition to those already defined using the filesystem convention in app/routes. Both sets of routes will be merged.

### ignoredRouteFiles

- **Optional**
- **Type**: `string[]`

https://remix.run/docs/en/v1/api/conventions#ignoredroutefiles

This is an array of globs (via minimatch) that Remix will match to files while reading your app/routes directory. If a file matches, it will be ignored rather that treated like a route module. This is useful for ignoring dotfiles (like .DS_Store files) or CSS/test files you wish to colocate.

## Usage

```js
import remixRoutes from "virtual:remix-routes";
```

**Example**

```jsx
import { render } from "react-dom";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import remixRoutes from "virtual:remix-routes";

// Note: This example only works with react-router >= 6.4
const router = createBrowserRouter([
  {
    path: "/",
    children: remixRoutes,
    // You can add your own `element`, `errorElement`, `loader`, ...
  },
]);

render(
  <RouterProvider router={router}>
    <App />
  </RouterProvider>,
  document.querySelector("#app")
);
```

**Example (`dataRouterCompatible = false`):**

```jsx
import { render } from "react-dom";
import { BrowserRouter, useRoutes } from "react-router-dom";
import remixRoutes from "virtual:remix-routes";

function App() {
  const element = useRoutes(remixRoutes);

  return <>{element}</>;
}

render(
  <BrowserRouter>
    <App />
  </BrowserRouter>,
  document.querySelector("#app")
);
```

## Async nested routes

When you configure routes to be imported asynchronously with the `importMode` option, it is important to note that this can create a request waterfall.

Lets say we land on the nested route `/one/two/three`.

React will first render (and load) the component for `one`, then `two` and at last `three` in series.\
Each component needs to be loaded and rendered before the next one is loaded.

This is not ideal so you can use the `EagerLoader` component exported by `vite-plugin-remix-routes/client` to immediately load all the components that will be needed for the current route.

**Example:**

```jsx
import { render } from "react-dom";
import { BrowserRouter, useRoutes } from "react-router-dom";
import { EagerLoader } from "vite-plugin-remix-routes/client";
import routes from "virtual:remix-routes";

function App() {
  const element = useRoutes(routes);

  return (
    <>
      <EagerLoader routes={routes} />
      {element}
    </>
  );
}

render(
  <BrowserRouter>
    <App />
  </BrowserRouter>,
  document.querySelector("#app")
);
```

Note that if you don't render an `<Outlet />` in one of the parent components, this will still load the subcomponent(s), even though React will not render it and would not have loaded it.\
But in that case, you probably don't want a nested route anyway.

<details>
<summary>How does this work?</summary>

[This is the code](https://github.com/vjee/vite-plugin-remix-routes/blob/main/lib/client/eager-loader.ts) for `EagerLoader`.
It gets the current location with the `useLocation` hook and gets all the matching routes for that location with `matchRoutes`.
Then we loop over each of the matching routes and call it's `loader` method.

This `loader` method is added to async routes by `vite-plugin-remix-routes` and looks like this: `loader: () => import("./path/to/route/component")`.

This will start the download of the route component. When React tries to render it later on, it is already loaded or it reuses the pending request if it hasn't finished yet.

</details>
<br />

More info about `useRoutes` can be found here:

- https://reactrouter.com/docs/en/v6/api#useroutes
- https://reactrouter.com/docs/en/v6/examples/route-objects

## TypeScript

If you use TypeScript you can add the following to your `vite-env.d.ts` file.\
This will add types for the `virtual:remix-routes` module.

```ts
/// <reference types="vite-plugin-remix-routes/virtual" />
```

## Similar projects

### vite-plugin-pages

This project is inspired by [`vite-plugin-pages`](https://github.com/hannoeru/vite-plugin-pages)
that can be used with both Vue and React.

`vite-plugin-remix-routes` is different in that it utilizes [`remix`](https://github.com/remix-run/remix) to generate the routes array instead of using a custom convention.\
As the name suggests, it also only works with React.

## License

[MIT](https://github.com/vjee/vite-plugin-remix-routes/blob/main/LICENSE)


CHANGELOG.md
# Changelog

## [0.4.1] - 2022-11-24

- Fix react peer dependency

## [0.4.0] - 2022-10-31

- Fix paths on windows environment

## [0.3.0] - 2022-10-25

- Dependency updates
- Add `dataRouterCompatible` option that defaults to `true` to be used with `react-router` 6.4 or later.
  - `importMode` option only works when `dataRouterCompatible` is `false`.
  - `is404Route` option only works when `dataRouterCompatible` is `false`.
- Custom `loader` property on routes with `importMode='async'` has been renamed to `importPromise` so it doesn't conflict with `react-router`'s new `loader` property.
  - `EagerLoader` has been updated to use `importPromise`.

## [0.2.0] - 2022-03-29

### Breaking

- `appDir` is now `appDirectory` to match remix's config
- `appDirectory` defaults to "app" instead of "src" to match remix

### Added

- Support remix's [`route`](https://remix.run/docs/en/v1/api/conventions#routes) and [`ignoredRouteFiles`](https://remix.run/docs/en/v1/api/conventions#ignoredroutefiles) configuration
- `build:watch` command for development

### Changed

- Tests now use vitest

## [0.1.3]

### Fixed

- Fix tests.

### Changed

- Use TypeScript for tests, scripts and examples.
- Update code block languages in README.md.

## [0.1.2]

### Changed

- Allow nested 404 routes.

## [0.1.1]

### Fixed

- Correctly stringify routes with `.` characters.

## [0.1.0]

- Initial version


lib/node/index.ts
import path from "node:path";
import fs from "node:fs";

import type { Plugin } from "vite";

import { getRoutes, RemixOptions } from "./remix";
import type { Route } from "./remix";
import { stringifyRoutes } from "./utils";

export interface Options extends PluginOptions, RemixOptions {}
export interface PluginOptions {
  /**
   * Set this to `false` if you're not using a router compatible with the data APIs
   * released in react-router 6.4.
   * https://reactrouter.com/en/main/routers/picking-a-router#using-v64-data-apis
   *
   * @default true
   */
  dataRouterCompatible?: boolean;

  /**
   * NOTE: This option only works if `dataRouterCompatible` is set to `false`.
   *
   * A function that receives a `Route` to determine if the route's component
   * should be imported synchronously or asynchronously.
   *
   * @default () => "sync"
   */
  importMode?: (route: Route) => "async" | "sync";

  /**
   * NOTE: This option only works if `dataRouterCompatible` is set to `false`. You should use an `ErrorBoundary` component instead.
   *
   * A function that receives a `Route` to determine if it should be a 404 route. (`path="*"`)
   * By default this matches the same 404 file as Remix does.
   * Keep in mind this only receives top level routes, so you can't mark nested routes
   * as 404 routes.
   *
   * @default (route) => route.id === "routes/404"
   */
  is404Route?: (route: Route) => boolean;
}

function plugin(options: Options = {}): Plugin {
  const virtualModuleId = "virtual:remix-routes";

  const {
    appDirectory = "app",
    dataRouterCompatible,
    importMode,
    is404Route,
    routes,
    ignoredRouteFiles,
  } = options;

  const dir = path.resolve(process.cwd(), appDirectory);
  const prefix = `.${path.sep}${path.relative(process.cwd(), dir)}`;

  if (
    !fs.existsSync(path.join(dir, "routes")) ||
    !fs.statSync(path.join(dir, "routes")).isDirectory()
  ) {
    throw new Error(
      `[vite-plugin-remix-routes] routes directory not found in appDirectory: ${path.relative(
        process.cwd(),
        appDirectory
      )}`
    );
  }


  return {
    name: "vite-plugin-remix-routes",

    resolveId(id) {
      if (id === virtualModuleId) {
        return id;
      }
    },

    async load(id) {
      if (id === virtualModuleId) {
        const generatedRoutes = await getRoutes({
          appDirectory: dir,
          dataRouterCompatible,
          is404Route,
          routes,
          ignoredRouteFiles,
        });

        const { routesString, componentsString } = stringifyRoutes(
          generatedRoutes,
          { prefix, dataRouterCompatible, importMode }
        );

        return (
          `import { createElement, lazy, useEffect } from 'react';\n` +
          `${componentsString}\n` +
          `export default ${routesString};\n`
        );
      }
    },
  };
}

export default plugin;
export { getRoutes, stringifyRoutes };

export type { Route };
